File: 16.最接近的三数之和.cpp
/*
 * @lc app=leetcode.cn id=16 lang=cpp
 *
 * [16] 最接近的三数之和
 */
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        std::sort(nums.begin(), nums.end());
        int minDiff = std::numeric_limits<int>::max();
        int minSum = 0;
        for(int i = 0;i < nums.size()-2;i++){
            if(i > 0 && nums[i] == nums[i-1])
                continue;
            int front = i+1;
            int tail = nums.size()-1;
            while(front < tail){
                if(front != i+1 && nums[front] == nums[front-1]){
                    front++;
                    continue;
                }
                if(tail != nums.size()-1 && nums[tail] == nums[tail+1]){
                    tail--;
                    continue;
                }
                int sum = nums[i]+nums[front]+nums[tail];
                int diff = abs(sum-target);
                if(diff == 0)
                    return sum;
                else{
                    if(minDiff > diff){
                        minDiff = diff;
                        minSum = sum;
                    }
                    if(sum-target > 0)
                        tail--;
                    else
                        front++;
                }
            }
        }
        return minSum;
    }
};

File: 16.最接近的三数之和.rs
/*
 * @lc app=leetcode.cn id=16 lang=rust
 *
 * [16] 最接近的三数之和
 */

// @lc code=start
// -5 -5 -4 0 0 3 3 4 5

impl Solution {
    pub fn three_sum_closest(mut nums: Vec<i32>, target: i32) -> i32 {
        nums.sort_unstable();
        let mut result: Option<i32> = None;
        for idx in 0..nums.len() {
            let mut head = idx + 1;
            let mut tail = nums.len() - 1;
            while head < tail {
                let t = nums[head] + nums[tail] + nums[idx];
                if let Some(result) = result.as_mut() {
                    if (target - t).abs() < (target - *result).abs() {
                        *result = t;
                    }
                } else {
                    result = Some(t);
                }
                if t > target {
                    tail -= 1;
                } else if t < target {
                    head += 1;
                } else {
                    return result.unwrap();
                }
            }
        }
        return result.unwrap();
    }
}
// @lc code=end

